borg/archiver.py

```diff
@@ -719,6 +719,107 @@ def do_upgrade(self, args):
+            nonlocal exit_soon
+            if exit_soon:
+                sys.stderr.write("Received signal, again. I'm not deaf.\n" )
+            else:
+                sys.stderr.write("Received signal, will exit cleanly.\n")
+            sys.stderr.flush()
+            exit_soon = True
+
+        signal.signal(signal.SIGTERM, do_exit_soon)
+        signal.signal(signal.SIGINT, do_exit_soon)
+
+        stats = Stats()
+        exit_soon = False
+
+        dry_run = args.dry_run
+        repository = self.open_repository(args, exclusive=True)
+        manifest, key = Manifest.load(repository)
+        compr_args = dict(buffer=COMPR_BUFFER)
+        compr_args.update(args.compression)
+        key.compressor = Compressor(**compr_args)
+        decompress_buffer = bytes(COMPR_BUFFER)
+
+        last_segment_id = repository.io.get_latest_segment()
+        segments = list(repository.io.segment_iterator())
+        progress_indicator = ProgressIndicatorPercent(len(segments), start=0, step=0.1, same_line=True,
+                                                      msg="%3.1f %% processed")
+
+        if not repository.io.is_committed_segment(segments[-1][1]):
+            self.print_error('Last segment in repository is uncomitted. Repository corrupted. Try '
+                             '"borg check --repair".')
+            repository.close()
+            return self.exit_code
+        for i, (segment_id, segment_filename) in enumerate(reversed(segments)):
+            chunks = {}
+            for tag, id_, offset, data in repository.io.iter_objects(segment_id, True):
+                if tag != TAG_PUT or set(id_) == {0}:
+                    continue
+                stats.chunks_seen += 1
+                stats.old_size += len(data)
+                decrypted_data = key.decrypt(id_, data, no_decompress=True)
+                compressor = detect_compressor(decrypted_data[:2])
+                if not args.force_recompress and isinstance(key.compressor.compressor, compressor):
+                    stats.chunks_skipped += 1
+                    stats.new_size += len(data)
+                    continue
+                decompressed_data = compressor(buffer=decompress_buffer).decompress(decrypted_data)
+                key.assert_chunk_id(id_, decompressed_data)
+                data = key.encrypt(decompressed_data)
+                stats.new_size += len(data)
+                chunks[id_] = data
+            if not dry_run:
+                try:
+                    for id_, data in chunks.items():
+                        repository.put(id_, data)
+                        stats.chunks_recompressed += 1
+                    if exit_soon or repository.io.get_latest_segment() > last_segment_id + 10:
+                        last_segment_id = repository.io.get_latest_segment()
+                        repository.commit()
+                    if exit_soon:
+                        break
+                except Exception as e:  # too broad!
```

Comment:
I added this part when evaluating "Ctrl-C safety" the first time (and it caught they KeyboardInterrupt reliably), but I think now that signals are handled explicitly it's unnecessary.